cluster.c

```diff
@@ -379,6 +379,58 @@ RRStatus ShardGroupAppendLogEntry(RedisRaftCtx *rr, ShardGroup *sg, int type, vo
     return RR_OK;
 }
 
+RRStatus ShardGroupsAppendLogEntry(RedisRaftCtx *rr, int num_sg, ShardGroup **sg, int type, void *user_data)
+{
+    /* Make sure we're still a leader, could have changed... */
+    /* SJP: unsure this is true in this case */
+    if (!raft_is_leader(rr->raft)) {
+        return RR_ERROR;
+    }
+
+    /* Serialize */
+    /* We reuse the existing shard group serialization
+     * Format:
+     * <num shard groups>:strlen(sg #1 payload):sg #1 pauload|...:strlen(sg #n payload):sg #n payload|
+     */
+    RedisModuleString * str = RedisModule_CreateStringPrintf(rr->ctx, "%d:", num_sg);
+    for (int i = 0; i < num_sg; i++) {
+        char *payload = ShardGroupSerialize(sg[i]);
+        RedisModuleString *payload_len = RedisModule_CreateStringPrintf(rr->ctx, "%lu:", strlen(payload));
+        size_t payload_len_str_len;
+        const char * payload_len_str = RedisModule_StringPtrLen(payload_len, &payload_len_str_len);
+        RedisModule_StringAppendBuffer(rr->ctx, str, payload_len_str, payload_len_str_len);
+        RedisModule_StringAppendBuffer(rr->ctx, str, payload, strlen(payload));
+        RedisModule_StringAppendBuffer(rr->ctx, str, ":", 1);
+        if (!payload) {
+            RedisModule_FreeString(rr->ctx, str);
+            return RR_ERROR;
+        }
+        RedisModule_Free(payload);
+    }
+
+    size_t len;
+    const char * payload = RedisModule_StringPtrLen(str, &len);
+
+    raft_entry_t *entry = raft_entry_new(len);
+    entry->type = type;
+    entry->id = rand();
+    entry->user_data = user_data;
+    memcpy(entry->data, payload, len);
+    RedisModule_Free(str);
```

Comment:
Would this make code a bit simpler ? 

```suggestion
    size_t buflen = 4096;
    char *buf = RedisModule_Alloc(buflen);

    buf = catsnprintf(buf, &buflen, "%d:", num_sg);
    for (int i = 0; i < num_sg; i++) {
        char *payload = ShardGroupSerialize(sg[i]);
        buf = catsnprintf(buf, &buflen, "%lu:%s:", strlen(payload), payload);
        RedisModule_Free(payload);
    }
    size_t payload_len = strlen(buf);
    raft_entry_t *entry = raft_entry_new(payload_len);
    entry->type = type;
    entry->id = rand();
    entry->user_data = user_data;
    memcpy(entry->data, buf, payload_len);
    RedisModule_Free(buf);
```